core/src/HIP/Kokkos_HIP_Team.hpp

```diff
@@ -922,6 +922,328 @@ KOKKOS_INLINE_FUNCTION void single(
 #endif
 }
 
+template <size_t RemainingRank>
+struct ParallelForMDTeamThreadRangeHIPImpl {
+ private:
+  template <typename Boundaries, typename Closure>
+  KOKKOS_INLINE_FUNCTION static void next_rank(
+      Boundaries const& boundaries, Closure const& closure,
+      typename Boundaries::index_type i) {
+    auto newClosure = [i, &closure](auto... is) { closure(i, is...); };
+    ParallelForMDTeamThreadRangeHIPImpl<RemainingRank - 1>::parallel_for_impl(
+        boundaries, newClosure);
+  }
+
+ public:
+  static constexpr size_t remaining_rank = RemainingRank;
+
+  template <typename Boundaries, typename Closure>
+  KOKKOS_INLINE_FUNCTION static void parallel_for_impl(
+      Boundaries const& boundaries, Closure const& closure) {
+    (void)boundaries;
+    (void)closure;
+
+#ifdef __HIP_DEVICE_COMPILE__
+    using index_type        = typename Boundaries::index_type;
+    using signed_index_type = std::make_signed_t<index_type>;
+
+    auto currentRank                  = Boundaries::rank - RemainingRank;
+    signed_index_type offsetThreadIdx = (currentRank) ? 0 : threadIdx.y;
+    signed_index_type offsetBlockDim  = (currentRank) ? 1 : blockDim.y;
```

Comment:
I guess this is good enough for a first implementation. But in the end, we probably want to use proper tiling here, right?